pkg/csi/service/fcd/controller.go

```diff
@@ -72,15 +75,133 @@ func (c *controller) CreateVolume(
+		ctx, datastoreName, datastoreType, volName, vclib.FindFCDByName)
+	if err == nil {
+		log.Warningf("Volume with name %s already exists. Checking for similar parameters.", volName)
+
+		if firstClassDisk.Config.CapacityInMB != volSizeMB {
+			return nil, status.Errorf(codes.Internal,
+				"Volume already exists but requesting different size. Existing %d != Requested %d",
+				firstClassDisk.Config.CapacityInMB, volSizeMB)
+		}
+	} else {
+		err = discoveryInfo.DataCenter.CreateFirstClassDisk(ctx, datastoreName, datastoreType, volName, volSizeMB)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "CreateFirstClassDisk failed. Err: %v", err)
+		}
+
+		firstClassDisk, err = discoveryInfo.DataCenter.GetFirstClassDisk(
+			ctx, datastoreName, datastoreType, volName, vclib.FindFCDByName)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "GetFirstClassDiskByName(%s) failed. Err: %v", volName, err)
+		}
+	}
+
+	attributes := make(map[string]string)
+	attributes[AttributeFirstClassDiskType] = FirstClassDiskTypeString
+	attributes[AttributeFirstClassDiskName] = firstClassDisk.Config.Name
+	attributes[AttributeFirstClassDiskParentType] = string(firstClassDisk.ParentType)
+	if firstClassDisk.ParentType == vclib.TypeDatastoreCluster {
+		attributes[AttributeFirstClassDiskParentName] = firstClassDisk.StoragePodInfo.Summary.Name
+		attributes[AttributeFirstClassDiskOwningDatastore] = firstClassDisk.DatastoreInfo.Info.Name
+	} else {
+		attributes[AttributeFirstClassDiskParentName] = firstClassDisk.DatastoreInfo.Info.Name
+	}
+
+	resp := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			VolumeId:      firstClassDisk.Config.Id.Id,
+			CapacityBytes: int64(units.FileSize(firstClassDisk.Config.CapacityInMB * MbInBytes)),
+			VolumeContext: attributes,
+			//TODO: ContentSource?
+		},
+	}
+
+	return resp, nil
 }
 
 func (c *controller) DeleteVolume(
 	ctx context.Context,
 	req *csi.DeleteVolumeRequest) (
 	*csi.DeleteVolumeResponse, error) {
 
-	return nil, nil
+	secrets := req.GetSecrets()
+
+	//check for required parameters
+	if secrets == nil {
+		return nil, status.Errorf(codes.Internal, "Delete secrets is a required parameter.")
+	} else if len(req.VolumeId) == 0 {
+		return nil, status.Errorf(codes.Internal, "Volume ID is a required parameter.")
+	} else if len(secrets[AttributeFirstClassDiskParentType]) == 0 {
+		return nil, status.Errorf(codes.Internal, "Volume parameter %s is a required parameter.", AttributeFirstClassDiskParentType)
+	} else if len(secrets[AttributeFirstClassDiskParentName]) == 0 {
+		return nil, status.Errorf(codes.Internal, "Volume parameter %s is a required parameter.", AttributeFirstClassDiskParentName)
+	}
+
+	// Volume Type
+	datastoreType := vclib.TypeDatastoreCluster
+	volType := secrets[AttributeFirstClassDiskParentType]
+	if volType == string(vclib.TypeDatastore) {
+		datastoreType = vclib.TypeDatastore
+	}
+
+	datastoreName := secrets[AttributeFirstClassDiskParentName]
```

Comment:
it may be the case that since we are only supporting a single datastore at this time, that this is a "fixed" parameter for the plugin (via config) for now. Definitely will have to think about this, but seems to be a limitation of the spec right now.

Reply:
using new method WhichVCandDCByFCDId